src/ast.rs

```diff
@@ -39,185 +45,12 @@ pub trait NodeVisitor<'a> {
     fn visit(&mut self, node: &'a dyn Node);
 }
 
-/**
- * Acceptor is implemented on Nodes which can be visited by a NodeVisitor.
- *
- * It generally invokes the visitor's visit() method on each of its children.
- *
- */
-pub trait Acceptor {
-    fn accept<'a>(&'a self, visitor: &mut dyn NodeVisitor<'a>);
-}
-
-impl<T: Acceptor> Acceptor for Option<T> {
-    fn accept<'a>(&'a self, visitor: &mut dyn NodeVisitor<'a>) {
-        if let Some(node) = self {
-            node.accept(visitor)
-        }
-    }
-}
-
-/// A helper trait to invoke the visit() or visit_mut() method on a field.
-trait VisitableField {
-    fn do_visit<'a>(&'a self, visitor: &mut dyn NodeVisitor<'a>);
-    fn do_visit_mut<V: NodeVisitorMut>(&mut self, visitor: &mut V) -> VisitResult;
-}
-
-impl<N: Node + NodeMut> VisitableField for N {
-    fn do_visit<'a>(&'a self, visitor: &mut dyn NodeVisitor<'a>) {
-        visitor.visit(self);
-    }
-
-    fn do_visit_mut<V: NodeVisitorMut>(&mut self, visitor: &mut V) -> VisitResult {
-        visitor.visit_mut(self)
-    }
-}
-
-impl<N: Node + NodeMut + CheckParse> VisitableField for Option<N> {
-    fn do_visit<'a>(&'a self, visitor: &mut dyn NodeVisitor<'a>) {
-        if let Some(node) = self {
-            node.do_visit(visitor);
-        }
-    }
-
-    fn do_visit_mut<V: NodeVisitorMut>(&mut self, visitor: &mut V) -> VisitResult {
-        visitor.visit_optional_mut(self)
-    }
-}
-
+#[derive(Copy, Clone, Debug)]
 pub struct MissingEndError {
     allowed_keywords: &'static [ParseKeyword],
     token: ParseToken,
 }
 
-pub type VisitResult = ControlFlow<MissingEndError>;
-
-/// Similar to NodeVisitor, but for mutable nodes.
-trait NodeVisitorMut {
-    /// will_visit (did_visit) is called before (after) a node's fields are visited.
-    fn will_visit_fields_of<N: NodeMut>(&mut self, node: &mut N);
-    fn visit_mut<N: NodeMut>(&mut self, node: &mut N) -> VisitResult;
-    fn did_visit_fields_of<'a, N: NodeMut>(&'a mut self, node: &'a mut N, flow: VisitResult);
-
-    // Visit an optional field, perhaps populating it.
-    fn visit_optional_mut<N: NodeMut + CheckParse>(&mut self, node: &mut Option<N>) -> VisitResult;
-}
-
-/**
- * A trait implemented on Nodes which can be visited by a NodeVisitorMut.
- *
- * It generally invokes the right visit() method on each of its children.
- */
-trait AcceptorMut {
-    fn accept_mut<V: NodeVisitorMut>(&mut self, visitor: &mut V);
-}
-
-impl<T: AcceptorMut> AcceptorMut for Option<T> {
-    fn accept_mut<V: NodeVisitorMut>(&mut self, visitor: &mut V) {
-        if let Some(node) = self {
-            node.accept_mut(visitor)
-        }
-    }
-}
-
-/// Node is the base trait of all AST nodes.
-pub trait Node: Acceptor + AsNode + std::fmt::Debug {
-    /// Return the kind of this node.
-    fn kind(&self) -> Kind<'_>;
-
-    /// Return the kind of this node, as a mutable reference.
-    fn kind_mut(&mut self) -> KindMut<'_>;
```

Comment:
7ecd2d0c594 (ast: move kind_mut back into AcceptorMut, 2025-09-03)
introduces `LIST_KIND` only to delete it later.  Should probably
remove it from history outright (to make the sum of deltas smaller).

Also, this seems to be doing multiple things at once:
- moving kind_mut() to AcceptorMut
- replacing empty list objects with LIST_KIND
- demoting JobList? (does that belong to a different one?)

Can it be split up to separate conerns?